test/georef_ocd_mapping_t.cpp

```diff
@@ -0,0 +1,261 @@
+	        << GeorefCoreData { 15000, 5.22, 5.22, { 52.39963782, 13.04811373 }, QStringLiteral("UTM"), { QStringLiteral("33") } };
+	QTest::newRow("Nasca airport - Peru, WGS 84/UTM zone 18S")
+	        << OcdGeorefFields { -2.70000000, 15000, 504606, 8358028, -2018, 1 }
+	        << GeorefCoreData { 15000, -2.70, -2.70, { -14.85213486, -74.95718575 }, QStringLiteral("UTM"), { QStringLiteral("18 S") } };
+	QTest::newRow("Wellington airport - New Zealand, WGS 84/UTM zone 60S")
+	        << OcdGeorefFields { 21.14000000, 15000, 316493, 5423682, -2060, 1 }
+	        << GeorefCoreData { 15000, 21.14, 21.14, { -41.31746197, 174.80754616 }, QStringLiteral("UTM"), { QStringLiteral("60 S") } };
+}
+
+}  // anonymous namespace
+
+class GeoreferenceMappingTest : public QObject
+{
+	Q_OBJECT
+
+private slots:
+	void initTestCase()
+	{
+		// empty
+	}
+	void testOcdToMapper_data()
+	{
+		commonTestData();
+
+		QTest::newRow("OCD CRS Local")
+		        << OcdGeorefFields { 15.35000000, 15000, 5556666, 6667777, 1000, 0 }
+		        << GeorefCoreData { 15000, 15.35, 15.35, { 0, 0 }, QStringLiteral("Local"), { } };
+		// exercise error paths
+		QTest::newRow("Unknown OCD grid/zone combination")
+		        << OcdGeorefFields { 15.35000000, 15000, 5556666, 6667777, 99999, 0 }
+		        << GeorefCoreData { 15000, 15.35, 15.35, { 0, 0 }, QStringLiteral("Local"), { } };
+	}
+
+	void testOcdToMapper()
+	{
+		QFETCH(OcdGeorefFields, ocd_string_fields);
+		QFETCH(GeorefCoreData, georef_result);
+
+		std::vector<QString> warnings;
+		auto add_warning = [&warnings](auto w) { warnings.push_back(w); };
+		Georeferencing georef;
+
+		ocd_string_fields.setupGeoref(georef, add_warning);
+
+		QStringList crs_params;
+		for (auto s : georef.getProjectedCRSParameters())
+			crs_params.push_back(s);
+
+		auto converted = GeorefCoreData {
+		                 georef.getScaleDenominator(),
+		                 georef.getDeclination(),
+		                 georef.getGrivation(),
+		                 georef.getGeographicRefPoint(),
+		                 georef.getProjectedCRSId(),
+		                 crs_params };
+
+		QCOMPARE(converted, georef_result);
+	}
+
+	void testMapperToOcd_data()
+	{
+		commonTestData();
+
+		QTest::newRow("Mapper CRS Local")
+		        << OcdGeorefFields { 21.32000000, 15000, 0, 0, 1000, 1 }
+		        << GeorefCoreData { 15000, 21.32, 21.32, { 58.972, -34.412 }, QStringLiteral("Local"), { } };
+		QTest::newRow("Mapper UTM 41N expressed as EPSG code")
+		        << OcdGeorefFields { 2.30000000, 10000, 285016, 5542944, 2041, 1 }
+		        << GeorefCoreData { 10000, 2.3, 2.3, { 50, 60 }, QStringLiteral("EPSG"), { QStringLiteral("32641") } };
+		QTest::newRow("Mapper UTM 21S expressed as EPSG code")
+		        << OcdGeorefFields { -2.70000000, 15000, 285016, 4457056, -2021, 1 }
+		        << GeorefCoreData { 15000, -2.70, -2.70, { -50, -60 }, QStringLiteral("EPSG"), { QStringLiteral("32721") } };
+		// exercise error paths
```

Comment:
Even in case of error, a, m, x, y need to be properly set. In addition, the test needs to verify that a warning was added (probably another test data field).

For verifying x and y, you must test with projected coordinates instead of geograpic coordinates. I think projected coordinates are more appropriate anyway: This is what is saved in the OCD file. And the test would no longer rely on proj transformations.

Reply:
> Even in case of error, a, m, x, y need to be properly set.

They are. See above.

> In addition, the test needs to verify that a warning was added (probably another test data field).

I had that in previous iterations but dropped the extra field. IMO it just cluttered the code while adding little value.

> For verifying x and y, you must test with projected coordinates instead of geographic coordinates.

I don't see why I exactly *must* switch to projected coords but the point about transformation is persuasive. I'll look into it.

Reply:
>> For verifying x and y, you must test with projected coordinates instead of geographic coordinates.
>
> I don't see why I exactly must switch to projected coords but the point about transformation is persuasive. I'll look into it.

With geographic coordinates, you bring PROJ.4 conversions into the code path under test. IMO this is conceptually wrong because it introduces complexity and more modes of error (PROJ.4 installation, additional floating point calculations) when we actually want to test a piece of code which does not deal (directly) with geographic coordinates. It is simply not this test's responsibility: Coordinate conversions with PRO.4 are to be covered by `GeoreferencingTest`.

The conversion to geographic coordinates also blocks the testing of the graceful (i.e. parameter-preserving) handling of unknown grids.

Reply:
ack

Reply:
> >    In addition, the test needs to verify that a warning was added (probably another test data field).
>
> I had that in previous iterations but dropped the extra field. IMO it just cluttered the code while adding little value.

So I gave the number_of_warnings test second chance and it's back. Updated patch set will follow shortly.